pkg/db/db_test.go

```diff
@@ -0,0 +1,194 @@
+	}
+	return tableNames, nil
+}
+
+// ensureNotTables returns a function that will fail a test if any of the provided table
+// names currently exists in the connected database.
+func ensureNotTables(unexpectedNames ...string) func(pg *sql.DB, t *testing.T) {
+	return func(pg *sql.DB, t *testing.T) {
+		names, err := tableNames(pg)
+		if err != nil {
+			t.Fatalf("couldn't list table rows: %v", err)
+		}
+		existing := make(map[string]bool)
+		for _, name := range names {
+			existing[name] = true
+		}
+		for _, name := range unexpectedNames {
+			if existing[name] {
+				t.Errorf("did not expect table %s to exist", name)
+			}
+		}
+	}
+}
+
+// ensureTables returns a function that will fail a test if any of the provided table names
+// currently does not exist in the connected database.
+func ensureTables(expectedNames ...string) func(pg *sql.DB, t *testing.T) {
+	return func(pg *sql.DB, t *testing.T) {
+		names, err := tableNames(pg)
+		if err != nil {
+			t.Fatalf("couldn't list table rows: %v", err)
+		}
+		existing := make(map[string]bool)
+		for _, name := range names {
+			existing[name] = true
+		}
+		for _, name := range expectedNames {
+			if !existing[name] {
+				t.Errorf("expected table %s to exist", name)
+			}
+		}
+	}
+}
+
+// migrationTestCase provides hooks for testing database migrations.
+// The `preup` function will be run before the up migration is applied to ensure
+// that state is as expected and to provide an opportunity to seed the database
+// with data.
+// the `during` function will be run after the up migration has been applied to
+// allow checking for the effects of the migration.
+// the `postdown` function will be run after the down migration has been applied
+// to ensure that the migration cleaned up after itself properly.
+type migrationTestCase struct {
+	preup, during, postdown func(pg *sql.DB, t *testing.T)
+}
+
+// migrationTests maps from migration numbers to the migrationTestCast that will
+// be used to check their correctness.
+var migrationTests = map[int]migrationTestCase{
+	1: {
+		preup:    ensureNotTables("jobs"),
+		during:   ensureTables("jobs"),
+		postdown: ensureNotTables("jobs"),
+	},
+	2: {
+		preup:    ensureNotTables("testcases"),
+		during:   ensureTables("testcases"),
+		postdown: ensureNotTables("testcases"),
+	},
+}
+
+// TestMigrations runs all configured migrations up and down, verifying their correctness
+// using the contents of `migrationTests`.
+func TestMigrations(t *testing.T) {
+	url := getDBURL(t)
+	if err := db.WithDB(url, func(pg *sql.DB) error {
+		return db.WithMigrator(pg, func(m *migrate.Migrate) error {
+			if err := m.Up(); err != nil {
+				t.Fatalf("Failed running all up migrations: %v", err)
+			}
+			maxVersion, _, err := m.Version()
+			if err != nil {
+				t.Fatalf("Did not expect error fetching final migration version: %v", err)
+			}
+			if err := m.Down(); err != nil {
+				t.Fatalf("Failed running all down migrations: %v", err)
+			}
```

Comment:
Is this testing a full-up then full-down.

That seems somewhat duplicative to the tests below, is there value in doing both?

Reply:
This was the only way I could find to determine how many migrations there are. I do this to get the value of `maxVersion`. Seems to me that our migrations should always be able to handing full up and then down like this, and the DB should be empty at this point so it should be pretty quick.

I considered writing the loop to just go until an error indicating that there weren't any more migrations. If you'd prefer that, I can try to reframe it that way.

Reply:
Nah, I'm good with that!